OutOfSchool/OutOfSchool.WebApi/Controllers/ChatWorkshopController.cs

```diff
@@ -56,30 +56,179 @@ public class ChatWorkshopController : ControllerBase
+        [ProducesResponseType(StatusCodes.Status401Unauthorized)]
+        [ProducesResponseType(StatusCodes.Status403Forbidden)]
+        [ProducesResponseType(StatusCodes.Status500InternalServerError)]
+        public Task<IActionResult> SetReadDatetimeInMessagesForParentAsync([Range(1, long.MaxValue)] long id)
+            => this.SetReadDatetimeInMessagesForCurrentUserAsync(id, this.IsParentAChatRoomParticipantAsync);
+
+        /// <summary>
+        /// Set current date and time for all not read chat messages for provider in the specified chat room.
+        /// </summary>
+        /// <param name="id">ChatRoom's Id.</param>
+        /// <returns>Number of successfully updated items.</returns>
+        [HttpPatch("provider/chatrooms/{id}/read")]
+        [Authorize(Roles = "provider")]
+        [ProducesResponseType(StatusCodes.Status200OK, Type = typeof(int))]
+        [ProducesResponseType(StatusCodes.Status204NoContent)]
+        [ProducesResponseType(StatusCodes.Status400BadRequest)]
+        [ProducesResponseType(StatusCodes.Status401Unauthorized)]
+        [ProducesResponseType(StatusCodes.Status403Forbidden)]
+        [ProducesResponseType(StatusCodes.Status500InternalServerError)]
+        public Task<IActionResult> SetReadDatetimeInMessagesForProviderAsync([Range(1, long.MaxValue)] long id)
+            => this.SetReadDatetimeInMessagesForCurrentUserAsync(id, this.IsProviderAChatRoomParticipantAsync);
+
+        private Task<bool> IsParentAChatRoomParticipantAsync(ChatRoomWorkshopDto chatRoom)
+        {
+            var userId = this.GetUserId();
+
+            return validationService.UserIsParentOwnerAsync(userId, chatRoom.ParentId);
+        }
+
+        private Task<bool> IsProviderAChatRoomParticipantAsync(ChatRoomWorkshopDto chatRoom)
+        {
+            var userId = this.GetUserId();
+
+            return validationService.UserIsWorkshopOwnerAsync(userId, chatRoom.WorkshopId);
+        }
+
+        private string GetUserId()
+        {
+            var userId = HttpContext.User.GetUserPropertyByClaimType(IdentityResourceClaimsTypes.Sub)
+                ?? throw new AuthenticationException($"Can not get user's claim {nameof(IdentityResourceClaimsTypes.Sub)} from HttpContext.");
+
+            return userId;
+        }
+
+        private Role GetUserRole()
+        {
+            var userRoleName = HttpContext.User.GetUserPropertyByClaimType(IdentityResourceClaimsTypes.Role)
+                ?? throw new AuthenticationException($"Can not get user's claim {nameof(IdentityResourceClaimsTypes.Sub)} from HttpContext.");
+
+            Role userRole = (Role)Enum.Parse(typeof(Role), userRoleName, true);
+
+            return userRole;
+        }
+
+        private void LogWarningAboutUsersTryingToGetNotOwnChatRoomInformation(long chatRoomId)
+        {
+            var messageToLog = $"{this.GetUserRole()} with UserId:{this.GetUserId()} is trying to get not his own chat room: {nameof(chatRoomId)}={chatRoomId}.";
```

Comment:
What if `GetUserRole` will throw an exception ?

Reply:
1. In each controller method, we catch AuthenticationException (that is thrown by this method) and log it, and return BadRequest to the user.
2. This is a helper method, so we use it after all these methods (GetUserRole and GetUserId) have been already used before to figure out if the user has rights. Here we only log warnings about inappropriate user behavior.